templates/boltzmann/core/honeycomb.ts

```diff
@@ -0,0 +1,1097 @@
+        // near future...
+        ignoreIncomingPaths,
+        ignoreOutgoingUrls
+      }),
+    ]
+
+    void `{% if redis %}`
+    is.push(new OtelRedisInstrumentation({}))
+    void `{% endif %}`
+
+    void `{% if postgres %}`
+    is.push(new OtelPgInstrumentation({}))
+    void `{% endif %}`
+
+    return is
+  },
+
+  // The SDK will take a service name, instrumentations
+  // and a trace exporter and give us a stateful singleton.
+  // This is that singleton!
+  sdk (
+    resource: otelResources.Resource,
+    instrumentations: OtelInstrumentation[]
+  ): OtelSDK {
+    return new OtelSDK({
+      resource,
+      instrumentations
+    })
+  }
+}
+
+// For testing purposes, it can be beneficial to
+// override how objects in OpenTelemetry initialization
+// are created. The Honeycomb class allows for
+// passing overrides into its constructor.
+interface OtelFactoryOverrides {
+  headers?: (writeKey: string | null, dataset: string | null) => HoneycombOTLPHeaders
+  resource?: (serviceName: string) => otelResources.Resource
+  tracerProvider?: (
+    resource: otelResources.Resource,
+  ) => NodeTracerProvider
+  spanExporter?: (protocol: string, headers: HoneycombOTLPHeaders) => otelTraceBase.SpanExporter
+  spanProcessor?: (spanExporter: otelTraceBase.SpanExporter) => otelTraceBase.SpanProcessor
+  instrumentations?: () => OtelInstrumentation[]
+  sdk?: (
+    resource: otelResources.Resource,
+    instrumentations: OtelInstrumentation[],
+  ) => OtelSDK;
+}
+
+// Let's GOOOOOOO
+class Honeycomb {
+  public options: HoneycombOptions
+  public features: HoneycombFeatures
+  public factories: OtelFactories
+
+  public tracerProvider: NodeTracerProvider | null
+  public spanExporter: otelTraceBase.SpanExporter | null
+  public spanProcessor: otelTraceBase.SpanProcessor | null
+  public instrumentations: OtelInstrumentation[] | null
+  public sdk: OtelSDK | null
+
+  public initialized: boolean
+  public started: boolean
+
+  private _logger: typeof bole | null
+
+  constructor(
+    options: HoneycombOptions,
+    overrides: OtelFactoryOverrides = {}
+  ) {
+    this.options = options
+    this.features = {
+      honeycomb: !options.disable,
+      beeline: !options.disable && !options.otel,
+      otel: options.otel || false
+    }
+    this.initialized = false
+    this.started = false
+
+    this.tracerProvider = null
+    this.spanExporter = null
+    this.spanProcessor = null
+    this.instrumentations = null
+    this.sdk = null
+
+    this.factories = {
+      ...defaultOtelFactories,
+      ...(overrides || {})
+    }
+
+    this._logger = null
+  }
+
+  get tracer (): otel.Tracer {
+    return otel.trace.getTracer('boltzmann', '{{ version }}')
+  }
+
+  // We (usually) load options from the environment. Unlike with Options,
+  // we do a lot of feature detection here.
+  public static fromEnv(
+    env: typeof process.env = process.env,
+    overrides: OtelFactoryOverrides = {}
```

Comment:
The *very* short answer is that I don't foresee boltzmann consumers overriding these. This overall structure is more motivated by unit testing than it is end users mucking about with it. That said, I can see some uses for implementers copying this file around.

Otherwise, we can talk about both the factories and the overrides in a follow-up zoom.

Reply:
I think we're on the same page now, with this being an internal-only API. I tried my best to make the comments more clear, but feel free to suggest different verbiage!